pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/BlockAwareSegmentInputStream.java

```diff
@@ -0,0 +1,219 @@
+    private int blockEntryCount;
+    // Number of payload Bytes read from ledger, and has been has been kept in this InputStream.
+    private int payloadBytesHave;
+    // Number of bytes that has been kept in this InputStream.
+    private int blockBytesHave;
+
+    // tracking read status for both header and entries.
+    // Bytes that already been read from this InputStream
+    private int bytesReadOffset = 0;
+    // Byte from this index is all padding byte
+    private int dataBlockFullOffset;
+    private final InputStream dataBlockHeaderStream;
+
+    // how many entries want to read from ReadHandle each time.
+    private static final int entriesNumberEachRead = 100;
+    // buf the entry size and entry id.
+    private static final int entryHeaderSize = 4 /* entry size*/ + 8 /* entry id */;
+    // Keep a list of all entries ByteBuf, each element contains 2 buf: entry header and entry content.
+    private List<CompositeByteBuf> entriesByteBuf = null;
+
+    public BlockAwareSegmentInputStream(ReadHandle ledger, long startEntryId, int blockSize) {
+        this.ledger = ledger;
+        this.startEntryId = startEntryId;
+        this.blockSize = blockSize;
+        this.dataBlockHeaderStream = DataBlockHeaderImpl.of(blockSize, startEntryId).toStream();
+        this.blockBytesHave = DataBlockHeaderImpl.getDataStartOffset();
+        this.payloadBytesHave = 0;
+        this.blockEntryCount = 0;
+        this.dataBlockFullOffset = blockSize;
+        this.entriesByteBuf = Lists.newLinkedList();
+    }
+
+    // read ledger entries.
+    private int readEntries() throws IOException {
+        checkState(bytesReadOffset >= DataBlockHeaderImpl.getDataStartOffset());
+        checkState(bytesReadOffset < dataBlockFullOffset);
+
+        try {
+            // once reach the end of entry buffer, start a new read.
+            if (entriesByteBuf.isEmpty()) {
+                readNextEntriesFromLedger();
+                log.debug("After readNextEntriesFromLedger: bytesReadOffset: {}, blockBytesHave: {}",
+                        bytesReadOffset, blockBytesHave);
+            }
+
+            // always read from the first ByteBuf in the list, once read all of its content remove it.
+            ByteBuf entryByteBuf = entriesByteBuf.get(0);
+            int ret = entryByteBuf.readByte();
+            bytesReadOffset ++;
+
+            if (entryByteBuf.readableBytes() == 0) {
+                entryByteBuf.release();
+                entriesByteBuf.remove(0);
+            }
+
+            return ret;
+        } catch (InterruptedException | ExecutionException e) {
+            log.error("Exception when get CompletableFuture<LedgerEntries>. ", e);
+            throw new IOException(e);
+        }
+    }
+
+    // read entries from ledger, and pre-handle the returned ledgerEntries.
+    private void readNextEntriesFromLedger() throws InterruptedException, ExecutionException {
```

Comment:
readNextEntriesFromLedger is doing too many things:
1. reads from ledger
2. updates cursor for where we've read to in ledger
3. updates number of entries written in this stream
4. updates number of bytes written in this stream
5. tracks whether there's enough space left in the block
6. updates end of block when there is not enough space

It should only do the first 2 as they concern reading. The rest of the actions are for "writing". In fact, I would make readNextEntriesFromLedger only do 1 and have readEntries deal with counting the entries.

So rather than having:
```
void readNextEntriesFromLedger()
```
have
```
List<ByteBuf> readNextEntries(long firstEntry, long maxNumberEntries)
```

Then in readEntries() increment blockEntryCount when we start reading from a new entry.

```
// once reach the end of entry buffer, start a new read.
if (entriesByteBuf.isEmpty()) {
    long firstEntryToRead = startEntryId + blockEntryCount;
    entriesByteBuf = readNextEntriesFromLedger(firstEntryToRead, ENTRIES_PER_READ);
    log.debug("After readNextEntriesFromLedger: bytesReadOffset: {}, blockBytesHave: {}",
              bytesReadOffset, blockBytesHave);
}

// always read from the first ByteBuf in the list, once read all of its content remove it.
ByteBuf entryByteBuf = entriesByteBuf.get(0);
int ret = entryByteBuf.readByte();

bytesReadOffset ++;

if (entryByteBuf.readableBytes() == 0) {
    entryByteBuf.release();
    entriesByteBuf.remove(0);

    blockEntryCount++;
}

return ret;
```
This way block entry count is being modified and used in the same method (i.e. you can see both uses in a single screen of code).

Methods with no parameters and a void return make code hard to understand as they are, by their very nature, full of side effects. If you want to understand a method that calls a zero-parameter void method, then you have to understand the whole zero-parameter void method also.

Similarly, we could move the handling for the padding into readEntries(), and have it all in one place.

```
// once reach the end of entry buffer, start a new read.
if (entriesByteBuf.isEmpty()) {
    long firstEntryToRead = startEntryId + blockEntryCount;
    entriesByteBuf = readNextEntriesFromLedger(firstEntryToRead, ENTRIES_PER_READ);
    log.debug("After readNextEntriesFromLedger: bytesReadOffset: {}, blockBytesHave: {}",
              bytesReadOffset, blockBytesHave);
}

// always read from the first ByteBuf in the list, once read all of its content remove it.
ByteBuf entryByteBuf = entriesByteBuf.get(0);
if ((bytesReadOffset + entryByteBuf.bytesAvailable()) <= blockSize) {
    int ret = entryByteBuf.readByte();
    bytesReadOffset ++;

    if (entryByteBuf.readableBytes() == 0) {
        entryByteBuf.release();
        entriesByteBuf.remove(0);
        blockEntryCount++;
    }
    return ret;
} else {
    return BLOCK_PADDING[bytesReadOffset++ % 4]; 
}
```

Again, it's mutation and usage in the same screen of text. And it removes the need for 2 member variable2 (dataBlockFullOffset & blockBytesHave).

Reply:
Thanks, changed it. 
First thought is to make readEntries handling easier, since it is called for each byte read; seems not too much compute added after change.